src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Instance.hs

```diff
@@ -28,75 +38,85 @@ import           ZkFold.Symbolic.Data.Eq
+
+instance (Arithmetic a, KnownNat n) => AdditiveMonoid (ArithmeticCircuit n a) where
+    zero = circuitN $ Vector <$> replicateM (Haskell.fromIntegral $ value @n) (newAssigned (const zero))
+
+instance (Arithmetic a, KnownNat n) => AdditiveGroup (ArithmeticCircuit n a) where
+    negate r = circuitN $ do
+        is <- runCircuit r
+        for is $ \i -> newAssigned (\x -> negate (x i))
+
+    r1 - r2 = circuitN $ do
+        is <- runCircuit r1
+        js <- runCircuit r2
+        for (Z.zip is js) $ \(i, j) -> newAssigned (\x -> x i - x j)
+
+instance (Arithmetic a, KnownNat n) => MultiplicativeSemigroup (ArithmeticCircuit n a) where
+    r1 * r2 = circuitN $ do
+        is <- runCircuit r1
+        js <- runCircuit r2
+        for (Z.zip is js) $ \(i, j) -> newAssigned (\x -> x i * x j)
+
+instance (Arithmetic a, KnownNat n) => Exponent (ArithmeticCircuit n a) Natural where
     (^) = natPow
 
-instance Arithmetic a => MultiplicativeMonoid (ArithmeticCircuit a) where
+instance (Arithmetic a, KnownNat n) => MultiplicativeMonoid (ArithmeticCircuit n a) where
     one = mempty
 
-instance (Arithmetic a, FromConstant b a) => FromConstant b (ArithmeticCircuit a) where
-    fromConstant c = embed (fromConstant c)
+-- TODO: The constant will be replicated in all outputs. Is this the desired behaviour?
+instance (Arithmetic a, FromConstant b a, KnownNat n) => FromConstant b (ArithmeticCircuit n a) where
+    fromConstant c = embedAll (fromConstant c)
 
-instance Arithmetic a => Semiring (ArithmeticCircuit a)
+instance (Arithmetic a, KnownNat n) => Semiring (ArithmeticCircuit n a)
 
-instance Arithmetic a => Ring (ArithmeticCircuit a)
+instance (Arithmetic a, KnownNat n) => Ring (ArithmeticCircuit n a)
 
-instance Arithmetic a => Exponent (ArithmeticCircuit a) Integer where
+instance (Arithmetic a, KnownNat n) => Exponent (ArithmeticCircuit n a) Integer where
     (^) = intPowF
 
-instance Arithmetic a => Field (ArithmeticCircuit a) where
+instance (Arithmetic a, KnownNat n) => Field (ArithmeticCircuit n a) where
     finv = invertC
-    rootOfUnity n = embed <$> rootOfUnity n
-
-instance Arithmetic a => BinaryExpansion (ArithmeticCircuit a) where
-    binaryExpansion r = circuits $ runCircuit r >>= expansion (numberOfBits @a)
-    fromBinary bits = circuit $ Haskell.traverse runCircuit bits >>= horner
-
-instance Arithmetic a => DiscreteField' (ArithmeticCircuit a) where
+    rootOfUnity n = embedAll <$> rootOfUnity n
+
+-- TODO: The only implementation that seems to make sense is when there is only one output.
+-- It is true?
+--
+-- Anyway, @binaryExpansion@ of an arithmetic circuit will return copies of the same circuit with different outputs.
+-- The whole point of this refactor was to avoid this.
+--
+-- Ideally, we want to return another ArithmeticCircuit with a number of outputs corresponding to the number of bits.
+-- This does not align well with the type of @binaryExpansion@
```

Comment:
I agree with this comment. Is it possible to change the signature of the `binaryExpansion` function? Maybe introduce a type family to handle both cases?

Reply:
+1, we also might want to have an `instance BinaryExpansion (UInt n a)`, but the current interface would force us to waste multitudes of registers

I propose we will redesign a `class BinaryExpansion` in one of our next PRs

Reply:
I think we had better look at the underlying algebra a little more carefully. For instance I believe that `BinaryExpansion` is really an internal encoding of the Euclidean degree function of a Euclidean semiring. 

> Ideally, we want to return another ArithmeticCircuit with a number of outputs corresponding to the number of bits.

I don't think this is possible is it? The number of bits isn't static; i.e. the binary expansion of `0` is `[]` but the binary expansion of `1` is `[1]`, etc.

Reply:
If we have `Type -> Type` parametrized outputs, then as an intermediate step you can have the binary expansion be a circuit `binaryExpansionC :: ArithmeticCircuit Identity a -> ArithmeticCircuit [] a`

Reply:
> I don't think this is possible if outputs are statically sized is it? 
In case of arithmetic circuits, the length of a list returned is always the same and is equal to `numberOfBits @a`

Reply:
Ok, that makes sense and I just tested it a little in GHCi.

Reply:
I still don't think that `Vector (NumberOfBits a)` is _needed_ but either it or `[]` can be used as wanted.